src/stan/mcmc/chainset.hpp

```
@@ -0,0 +1,556 @@
+    }
+    return thinned_samples;
+  }
+
+  static bool is_valid(const stan::io::stan_csv& stan_csv) {
+    if (stan_csv.header.empty())
+      return false;
+    if (stan_csv.samples.size() == 0)
+      return false;
+    if (stan_csv.samples.rows() != thinned_samples(stan_csv))
+      return false;
+    return true;
+  }
+
+  /**
+   * Process first chain: record header, thinned samples,
+   * add samples to vector chains.
+   */
+  void init_from_stan_csv(const stan::io::stan_csv& stan_csv) {
+    if (!is_valid(stan_csv)) {
+      throw std::invalid_argument("Invalid sample");
+    }
+    if (chains_.size() > 0) {
+      throw std::invalid_argument("Cannot re-initialize chains object");
+    }
+    param_names_ = stan_csv.header;
+    num_samples_ = thinned_samples(stan_csv);
+    chains_.push_back(stan_csv.samples);
+  }
+
+  /**
+   * Process next chain: validate size, shape, column names,
+   * append to vector chains.
+   */
+  void add(const stan::io::stan_csv& stan_csv) {
+    if (!is_valid(stan_csv)) {
+      throw std::invalid_argument("Invalid sample");
+    }
+    if (stan_csv.header.size() != num_params()) {
+      throw std::invalid_argument(
+          "Error add(stan_csv): number of columns in"
+          " sample does not match first chain");
+    }
+    if (thinned_samples(stan_csv) != num_samples_) {
+      throw std::invalid_argument(
+          "Error add(stan_csv): number of sampling draws in"
+          " sample does not match first chain");
+    }
+    for (int i = 0; i < num_params(); i++) {
+      if (param_names_[i] != stan_csv.header[i]) {
+        std::stringstream ss;
+        ss << "Error add(stan_csv): header " << param_names_[i]
+           << " does not match chain's header (" << stan_csv.header[i] << ")";
+        throw std::invalid_argument(ss.str());
+      }
+    }
+    chains_.push_back(stan_csv.samples);
+  }
+
+ public:
+  explicit chainset(const stan::io::stan_csv& stan_csv) {
+    init_from_stan_csv(stan_csv);
+  }
```

I would just use an Member initializer list. I think it's reasonable to assume the user will bring in a valid `stan_csv` so idt this function should have to check that

```suggestion
  explicit chainset(const stan::io::stan_csv& stan_csv) : 
    param_names_(stan_csv.header),
    num_samples_(thinned_samples(stan_csv)),
    chains_(stan_csv.samples) {}
```

---

I think it's safer to check - class `stan::io::stan_csv_reader` doesn't do these checks and they're not particularly expensive.

---

this isn't a very expensive check - but maybe it should throw an error instead of returning true/false.